### PR TITLE
Fixing the Android Manifest file

### PR DESCRIPTION
**Pull request by amanj120**

### CODE REVIEW

File: platform/android/export/export.h [hunk 1]
```diff
@@ -31,6 +31,82 @@
 #ifndef ANDROID_EXPORT_H
 #define ANDROID_EXPORT_H
 
+#include <core/ustring.h>
+
 void register_android_exporter();
 
+String ANDROID_MANIFEST_TEXT = "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n"
```
Comment (thanwah) [inline platform/android/export/export.h]: Add a TODO saying we want to find a better place for this, e.g. a templates directory

File: platform/android/export/export.h [hunk 2]
```diff
@@ -31,6 +31,82 @@
...
+                               "<manifest xmlns:android=\"http://schemas.android.com/apk/res/android\"\n"
+                               "    xmlns:tools=\"http://schemas.android.com/tools\"\n"
+                               "    package=\"PACKAGE_NAME_HERE\"\n"
```
Comment (thanwah) [inline platform/android/export/export.h]: Did you consider something like the `*PACKAGE_NAME*` style?

File: platform/android/export/export.h (hunk 1 above)
Comment (thanwah) [inline platform/android/export/export.h]: Move this to the cpp file and make it const

File: platform/android/export/export.cpp [hunk 3]
```diff
@@ -846,30 +912,7 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
-			if (!user_perm.empty()) {
-				perms.push_back(user_perm);
-			}
-		}
-
-		if (p_give_internet) {
-			if (perms.find("android.permission.INTERNET") == -1)
-				perms.push_back("android.permission.INTERNET");
-		}
+		Vector<String> perms = _get_permissions(p_preset, p_give_internet);
```
Comment (thanwah) [inline platform/android/export/export.cpp]: Please make a PR that precedes this PR that is focused on refactoring (extracting) this into a new function

File: platform/android/export/export.cpp [hunk 4]
```diff
@@ -795,8 +795,74 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+        bool min_gles3 = ProjectSettings::get_singleton()->get("rendering/quality/driver/driver_name") == "GLES3" &&
+                         !ProjectSettings::get_singleton()->get("rendering/quality/driver/fallback_to_gles2");
+        String gles_version = min_gles3 ? "0x00030000" : "0x00020000";
+        manifest_text = manifest_text.replace("GLES_VERSION_HERE", gles_version);
+
+		int orientation = p_preset->get("screen/orientation");
+		manifest_text = manifest_text.replace("GLES_VERSION_HERE", itos(orientation));
+
+
+        String screen_support_small = p_preset->get("screen/support_small") ? "true" : "false";
```
Comment (thanwah) [inline platform/android/export/export.cpp]: Always run clang-format as one of the steps for your PR checklist

**Review by thanwah — CHANGES_REQUESTED** [review]

File: platform/android/export/export.cpp (hunk 3 above)
Comment (amanj120) [reply platform/android/export/export.cpp]: Not sure why this is still showing up even after i merged the permissions branch with master

File: platform/android/export/export.cpp [hunk 5]
```diff
@@ -815,11 +891,93 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+
+		bool min_gles3 = ProjectSettings::get_singleton()->get("rendering/quality/driver/driver_name") == "GLES3" &&
+						 !ProjectSettings::get_singleton()->get("rendering/quality/driver/fallback_to_gles2");
+		String gles_version = min_gles3 ? "0x00030000" : "0x00020000";
+		manifest_text = manifest_text.replace("*GLES_VERSION*", gles_version);
+
+		int orientation = p_preset->get("screen/orientation"); //TODO: why is this an int?
+		manifest_text = manifest_text.replace("*SCREEN_ORIENTATION*", itos(orientation));
+
+		String screen_support_small = p_preset->get("screen/support_small") ? "true" : "false";
```
Comment (thanwah) [inline platform/android/export/export.cpp]: Maybe add a little helper function?

```
String _get_boolean_string(bool value) {
  return value ? "true" : "false";
}
```

to use here and below?

File: platform/android/export/export.cpp [hunk 6]
```diff
@@ -923,29 +924,35 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+		_get_permissions(p_preset, p_give_internet, &perms);
 		String permission_string;
 		for (int i = 0; i < perms.size(); i++) {
-			permission_string.insert(0, "\t<uses-permission android:name=\"" + perms.get(i) + "\"/>\n");
+			permission_string = permission_string.insert(0, "\t<uses-permission android:name=\"" + perms.get(i) + "\"/>\n");
 		}
 		manifest_text = manifest_text.replace("*PERMISSIONS*", permission_string);
 
 		int xr_mode_index = p_preset->get("xr_features/xr_mode");
 		String focus_awareness = p_preset->get("xr_features/focus_awareness") ? "true" : "false";
```
Comment (thanwah) [inline platform/android/export/export.cpp]: Also use bool_to_string() here

File: platform/android/export/export.cpp [hunk 7]
```diff
@@ -198,6 +198,84 @@ static const char *android_perms[] = {
...
+									 "    <!-- Any tag in this line after android:icon will be erased when doing custom builds. -->\n"
+									 "    <!-- If you want to add tags manually, do before it. -->\n"
+									 "    <!-- WARNING: This should stay on a single line until the parsing code is improved. See GH-32414. -->\n"
+									 "    <application android:label=\"@string/godot_project_name_string\" android:allowBackup=\"false\" tools:ignore=\"GoogleAppIndexingWarning\" android:icon=\"@mipmap/icon\" >\n"
+									 "\n"
+									 "        <!-- The following metadata values are replaced when Godot exports, modifying them here has no effect. -->\n"
+									 "        <!-- Do these changes in the export preset. Adding new ones is fine. -->\n"
+									 "\n"
+									 "        <!-- XR mode metadata. This is modified by the exporter based on the selected xr mode. DO NOT CHANGE the values here. -->\n"
+									 "        <meta-data\n" //
```
Comment (thanwah) [inline platform/android/export/export.cpp]: So in the case that XR isn't needed, we'll have something like `android:name=\"*XR_MODE_METADATA_NAME*\"` unchanged in the final manifest?

Instead should this section look like this:
`XR_MODE_PLACEHOLDER`

And then if there's nothing to include we replace the placeholder with a blank line, but if there is something we replace it with the whole `<meta-data` element?

Comment (amanj120) [reply platform/android/export/export.cpp]: Should I do the same thing with focus awareness?

File: platform/android/export/export.cpp [hunk 8]
```diff
@@ -198,6 +198,80 @@ static const char *android_perms[] = {
 	NULL
 };
 
+//TODO: put this somewhere better (i.e. templates folder)
+const String ANDROID_MANIFEST_TEXT = "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n"
+									 "<manifest xmlns:android=\"http://schemas.android.com/apk/res/android\"\n"
```
Comment (amanj120) [inline platform/android/export/export.cpp]: Running clang formatting gives me this super huge spacing

Comment (thanwah) [reply platform/android/export/export.cpp]: What if line 202 looks like this? Same?

```
const String ANDROID_MANIFEST_TEXT =
  "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n"
```

File: platform/android/export/export.cpp [hunk 9]
```diff
@@ -198,6 +198,80 @@ static const char *android_perms[] = {
...
+									 "\n"
+									 "    <!-- Any tag in this line after android:icon will be erased when doing custom builds. -->\n"
+									 "    <!-- If you want to add tags manually, do before it. -->\n"
+									 "    <!-- WARNING: This should stay on a single line until the parsing code is improved. See GH-32414. -->\n"
+									 "    <application android:label=\"@string/godot_project_name_string\" android:allowBackup=\"false\" tools:ignore=\"GoogleAppIndexingWarning\" android:icon=\"@mipmap/icon\" >\n"
+									 "\n"
+									 "        <!-- The following metadata values are replaced when Godot exports, modifying them here has no effect. -->\n"
+									 "        <!-- Do these changes in the export preset. Adding new ones is fine. -->\n"
+									 "\n"
+									 "        <!-- XR mode metadata. This is modified by the exporter based on the selected xr mode. DO NOT CHANGE the values here. -->\n"
```
Comment (thanwah) [inline platform/android/export/export.cpp]: This comment doesn't seem to be valid anymore. Remove?

File: platform/android/export/export.cpp [hunk 10]
```diff
@@ -795,31 +869,125 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
...
+
+		int xr_mode_index = p_preset->get("xr_features/xr_mode");
+		String focus_awareness = bool_to_string(p_preset->get("xr_features/focus_awareness"));
+		String plugins_names = get_plugins_names(get_enabled_plugins(p_preset));
+		if (plugins_names.empty()) {
+			plugins_names = "plugins_value";
+		}
+		String feature_string;
+
+		if (xr_mode_index == 1 /* XRMode.OVR */) {
```
Comment (thanwah) [inline platform/android/export/export.cpp:959]: There should be code in the else case something like this:

`manifest_text = manifest_text.replace("*XR_MODE_TAG*", "");`

**Review by thanwah — APPROVED** [review]

**Pull request merged**